File: examples/go/handlers/ollama.go
package handlers

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

const (
	ollamaBaseURL      = "http://localhost:11434"
	defaultOllamaModel = "qwen2.5:3b-instruct"
)

// ── /api/chat wire types ──────────────────────────────────────────────────────

type ollamaChatPayload struct {
	Model  string `json:"model"`  // default: qwen2.5:3b-instruct
	Prompt string `json:"prompt"` // user message content
}

type ollamaChatMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// ollamaChatRequest is the body sent to POST /api/chat.
type ollamaChatRequest struct {
	Model    string              `json:"model"`
	Messages []ollamaChatMessage `json:"messages"`
	Stream   bool                `json:"stream"` // always false
}

// ollamaChatAPIResponse is the relevant subset of Ollama's /api/chat response.
type ollamaChatAPIResponse struct {
	Model   string            `json:"model"`
	Message ollamaChatMessage `json:"message"`
	Done    bool              `json:"done"`
}

// ollamaChatResult is written to the output file and returned as the task result.
type ollamaChatResult struct {
	Model           string `json:"model"`
	Prompt          string `json:"prompt"`
	Response        string `json:"response"`
	ResponsePreview string `json:"response_preview"`
	ElapsedMs       int64  `json:"elapsed_ms"`
	OutputFile      string `json:"output_file"`
}

// OllamaChat sends a user message to a locally running Ollama instance via
// /api/chat and writes the response to examples/output/.
//
// Payload:  {"model": "qwen2.5:3b-instruct", "prompt": "What's the weather like today?"}
// Output:   examples/output/go_ollama_<timestamp>.txt
//
// If Ollama is not running the task returns an error, triggering normal
// retry / dead logic. Start Ollama with: ollama serve
func OllamaChat(ctx context.Context, payload []byte) ([]byte, error) {
	var p ollamaChatPayload
	if err := json.Unmarshal(payload, &p); err != nil {
		return nil, fmt.Errorf("invalid payload: %w", err)
	}
	if p.Prompt == "" {
		return nil, fmt.Errorf("prompt is required")
	}
	if p.Model == "" {
		p.Model = defaultOllamaModel
	}

	reqBody, err := json.Marshal(ollamaChatRequest{
		Model: p.Model,
		Messages: []ollamaChatMessage{
			{Role: "user", Content: p.Prompt},
		},
		Stream: false,
	})
	if err != nil {
		return nil, fmt.Errorf("marshal request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost,
		ollamaBaseURL+"/api/chat", bytes.NewReader(reqBody))
	if err != nil {
		return nil, fmt.Errorf("build request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: 2 * time.Minute}
	start := time.Now()
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf(
			"Ollama unavailable at %s (run: ollama serve): %w",
			ollamaBaseURL, err,
		)
	}
	defer resp.Body.Close()
	elapsed := time.Since(start).Milliseconds()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("Ollama returned HTTP %d", resp.StatusCode)
	}

	var ollamaResp ollamaChatAPIResponse
	if err := json.NewDecoder(resp.Body).Decode(&ollamaResp); err != nil {
		return nil, fmt.Errorf("decode response: %w", err)
	}

	responseText := ollamaResp.Message.Content
	preview := responseText
	if len(preview) > 200 {
		preview = preview[:200] + "…"
	}

	logContent := fmt.Sprintf(
		"=== go_ollama_chat ===\nmodel:      %s\nprompt:     %s\nelapsed_ms: %d\n\n--- response ---\n%s\n",
		p.Model, p.Prompt, elapsed, responseText,
	)
	outPath, err := writeOutputText(
		fmt.Sprintf("go_ollama_%d.txt", time.Now().UnixNano()),
		logContent,
	)
	if err != nil {
		return nil, err
	}

	return json.Marshal(ollamaChatResult{
		Model:           ollamaResp.Model,
		Prompt:          p.Prompt,
		Response:        responseText,
		ResponsePreview: preview,
		ElapsedMs:       elapsed,
		OutputFile:      outPath,
	})
}

File: examples/python/handlers/ollama.py
"""py_ollama_chat — call a locally running Ollama instance via /api/chat."""
from __future__ import annotations

import logging
import time
from pathlib import Path

import httpx

OLLAMA_BASE     = "http://localhost:11434"
DEFAULT_MODEL   = "qwen2.5:3b-instruct"
OUTPUT_DIR      = Path(__file__).parent.parent.parent / "output"
logger          = logging.getLogger(__name__)


def handle_ollama_chat(task_id: str, payload: dict) -> dict:
    """
    Send a user message to local Ollama via /api/chat and save the response.

    Payload: {"model": "qwen2.5:3b-instruct", "prompt": "What's the weather like today?"}
    Output:  examples/output/py_ollama_<task_id>.txt

    If Ollama is not running the handler raises an exception, which causes
    ErmineTQ to retry the task according to the configured retry budget.
    Start Ollama with: ollama serve
    Pull model with:   ollama pull qwen2.5:3b-instruct
    """
    prompt = payload.get("prompt")
    if not prompt:
        raise ValueError("prompt is required")
    model = payload.get("model", DEFAULT_MODEL)

    request_body = {
        "model":   model,
        "messages": [{"role": "user", "content": prompt}],
        "stream":  False,
    }

    logger.info("calling Ollama  model=%s  prompt=%r", model, prompt[:80])
    start = time.monotonic()

    try:
        resp = httpx.post(
            f"{OLLAMA_BASE}/api/chat",
            json=request_body,
            timeout=120,
        )
        resp.raise_for_status()
    except httpx.ConnectError as exc:
        raise RuntimeError(
            f"Ollama not available at {OLLAMA_BASE}. "
            "Start it with: ollama serve"
        ) from exc

    elapsed_ms = int((time.monotonic() - start) * 1000)
    data = resp.json()

    response_text = data.get("message", {}).get("content", "")

    log_content = (
        f"=== py_ollama_chat ===\n"
        f"model:      {model}\n"
        f"prompt:     {prompt}\n"
        f"elapsed_ms: {elapsed_ms}\n\n"
        f"--- response ---\n{response_text}\n"
    )

    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    out_path = OUTPUT_DIR / f"py_ollama_{task_id}.txt"
    out_path.write_text(log_content, encoding="utf-8")

    preview = response_text[:200] + ("…" if len(response_text) > 200 else "")
    logger.info("response saved → %s  (%dms)", out_path, elapsed_ms)

    return {
        "model":            model,
        "response_preview": preview,
        "elapsed_ms":       elapsed_ms,
        "output_file":      str(out_path),
    }
